src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -779,6 +779,13 @@ module Crystal
     def type_assign(target : InstanceVar, value, node)
       # Check if this is an instance variable initializer
       unless @scope
+        # `InstanceVar` assignment appered in block is not checked
+        # by `Crystal::InstanceVarsInitializerVisitor` because this block
+        # may be passed to a macro. So, it checks here.
+        if current_type.is_a?(Program) || current_type.is_a?(FileModule)
```

Comment:
Apologize for the late reply. I missed your comment, sorry...

For example, `InstanceVarsInitializerVisitor` treats only `Program` and `FileModule` as global type. ([here](https://github.com/crystal-lang/crystal/blob/a82ef1b9229a90b2b5035ee30977f612f480a772/src/compiler/crystal/semantic/instance_vars_initializer_visitor.cr#L63))
Perhaps your suggestion is correct, but I don't know really.